### ISSUE TITLE
[RFE] trace exec - K8s enrichment - Pod label & annotations

### ISSUE THREAD

**Issue by galofir-ms** [inspektor-gadget/inspektor-gadget#2318]
Currently K8s container enrichment does not add pod labels and annotations to k8s context. We would like to add these fields as to further provide visibility for our customers and future security features that we are planning to implement which will depend on these fields.

**Comment by burak-ok** [comment]
Initial Podlabel enrichment for `ig` was merged with https://github.com/inspektor-gadget/inspektor-gadget/pull/2398

**Comment by blanquicet** [comment]
Making it generic for all gadgets, not just `trace_exec`.

**Comment by blanquicet** [comment]
We should consider using the API server to get this information as we did for the owner reference in https://github.com/inspektor-gadget/inspektor-gadget/pull/3360.

**Comment by burak-ok** [comment]
> For `ig`, we should consider using the API server to get this information as we did for the owner reference in #3360.

What would the advantage be to use the API server? I think having the information available without going through the API server is even better.
Of course without the API server we need `--runtime-protocol=cri`.

**Comment by alban** [comment]
> > For `ig`, we should consider using the API server to get this information as we did for the owner reference in #3360.
> 
> What would the advantage be to use the API server? I think having the information available without going through the API server is even better. Of course without the API server we need `--runtime-protocol=cri`.

Pod annotations can be modified during the lifetime of the pod. When getting the annotations from the cri, they are never updated. If we get them from the apiserver, we get the updates. I am not so convinced that getting updates for annotations is useful though.

**Comment by burak-ok** [comment]
> When getting the annotations from the cri, they are never updated.

That is true that we could periodically update them in the background. Would be periodic since contacting the API server for every event would be way too much overhead

**Comment by blanquicet** [comment]
I mean, I'm not saying to not retrieve these values from the container runtime, but get them all from both. It would make IG easier to use and understand. I noticed it when trying to write the documentation:

### Current state

| Flags | Owner reference | Pod labels | Annotations |
|---|---|---|---|
| `enrich-with-k8s-apiserver` | ✅ | ❌ | ❌ |
| `runtime-protocol=cri` | ❌ | ✅ | ❌ |

### My proposal

> **NOTE:** Events will only get enriched with the owner reference when using the Kubernetes API server because such a information is not available at the container runtimes.

| Flags | Owner reference | Pod labels | Annotations |
|---|---|---|---|
| `enrich-with-k8s-apiserver` | ✅ | ✅ | ✅ |
| `runtime-protocol=cri` | ❌ | ✅ | ✅ |

It will also allow to easily state that `kubectl-gadget` and `ig` enrich events with the same Kubernetes information when `ig` is used with `enrich-with-k8s-apiserver`. If users don't want to depend on the availability of the API server, then the events will not be enriched with the owner reference.

cc @mauriciovasquezbernal as we were discussing about it in https://github.com/inspektor-gadget/inspektor-gadget/pull/3430#pullrequestreview-2292825444.

**Comment by burak-ok** [comment]
> I mean, I'm not saying to not retrieve these values from the container runtime, but get them all from both.

Of course, that makes sense.

> It will also allow to easily state that kubectl-gadget and ig enrich events with the same Kubernetes information when ig is used with the `enrich-with-k8s-apiserver` flag

The flag `enrich-with-k8s-apiserver` does not enable Operators that rely on the KubeAPIServer. Created an issue to add that in https://github.com/inspektor-gadget/inspektor-gadget/issues/3484

**Comment by blanquicet** [comment]
To summarize the task in this issue:

- [ ] `enrich-with-k8s-apiserver` must also enrich events with podlabels and annotations
- [ ] `runtime-protocol=cri` must also enrich with annotations.

**Comment by stasyu2009-ux** [comment]
Всім дуже дякую
